README.md
# UTM dart  
Lat Lon and UTM bidirectional converter.  
translate from [Python UTM library](https://github.com/Turbo87/utm)  
  
Support Geodetic System  
- WGS84 (default)
- GRS80
- Bessel

The UTM coordinate system is explained on this [Wikipedia](https://en.wikipedia.org/wiki/Universal_Transverse_Mercator_coordinate_system) page.

## Usage

Simple usage.  
Default geodetic system is WGS84
```dart
import 'package:utm/utm.dart';

final utm = UTM.fromLatLon(lat: -30, lon: -150);
print('zone: ${utm.zone}');
print('N: ${utm.northing}');
print('E: ${utm.easting}');
print('lat: ${utm.lat}');
print('lat: ${utm.lon}');

final latlon = UTM.fromUtm(
  easting: utm.easting,
  northing: utm.northing,
  zoneNumber: utm.zoneNumber,
  zoneLetter: utm.zoneLetter,
);
print('lat: ${latlon.lat}');
print('lon: ${latlon.lon}');

```
  
Use specific geodetic system  
```dart
UTM.fromLatLon(lat: -30, lon: -150, type: GeodeticSystemType.bessel);
UTM.fromUtm(
  easting: utm.easting,
  northing: utm.northing,
  zoneNumber: utm.zoneNumber,
  zoneLetter: utm.zoneLetter,
  type: GeodeticSystemType.grs80,
);
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: 


CHANGELOG.md
# Changelog

## 0.0.1

- Initial version, translate from Python library

## 0.0.2

- Formatting and add comments

## 0.0.3

- add description in pubspec.yaml

## 0.1.0

- add arguments validator

## 0.1.1

- down grade meta package. because Flutter dependency version 1.1.6 meta package

## 0.1.2

- export UtmCoordinate

## 0.1.3

- migrate to null-safety


test/utm_test.dart
import 'package:utm/utm.dart';
import 'package:test/test.dart';

void main() {
  test('latlon to utm Test in tokyo', () {
    final utm = UTM.fromLatLon(lat: 35.720122, lon: 139.752899);
    expect(utm.northing, closeTo(3953623.37, 0.1));
    expect(utm.easting, closeTo(387203.09, 0.1));
    expect(utm.zoneLetter, equals('S'));
    expect(utm.zoneNumber, equals(54));
  });
  test('utm to lat lon Test in tokyo', () {
    final utm = UTM.fromUtm(
        easting: 387203.09,
        northing: 3953623.37,
        zoneNumber: 54,
        zoneLetter: 'S');
    expect(utm.northing, closeTo(3953623.37, 0.1));
    expect(utm.easting, closeTo(387203.09, 0.1));
    expect(utm.zone, equals('54S'));
    expect(utm.lat, closeTo(35.720122, 0.1));
    expect(utm.lon, closeTo(139.752899, 0.1));
  });
  test('latlon to utm Test in Sydney', () {
    final utm = UTM.fromLatLon(lat: -34.452211, lon: 150.947302);
    expect(utm.northing, closeTo(6185790.53, 0.1));
    expect(utm.easting, closeTo(311433.56, 0.1));
    expect(utm.zoneLetter, equals('H'));
    expect(utm.zoneNumber, equals(56));
  });
  test('latlon to utm Test in zone 32', () {
    final utm = UTM.fromLatLon(lat: 63.506144, lon: 9.20091);
    expect(utm.northing, closeTo(7042000, 0.1));
    expect(utm.easting, closeTo(510000, 0.1));
    expect(utm.zoneLetter, equals('V'));
    expect(utm.zoneNumber, equals(32));
  });
  test('latlon to utm Test in zone 33', () {
    final utm = UTM.fromLatLon(lat: 72.506144, lon: 20.20091);
    expect(utm.zoneLetter, equals('X'));
    expect(utm.zoneNumber, equals(33));
  });
  test('latlon to utm Test in zone 35', () {
    final utm = UTM.fromLatLon(lat: 72.506144, lon: 30.20091);
    expect(utm.zoneLetter, equals('X'));
    expect(utm.zoneNumber, equals(35));
  });
  test('latlon to utm Test in zone 37', () {
    final utm = UTM.fromLatLon(lat: 72.506144, lon: 40.20091);
    expect(utm.zoneLetter, equals('X'));
    expect(utm.zoneNumber, equals(37));
  });
}


test/validate_test.dart
import 'package:test/test.dart';
import 'package:utm/utm.dart';

void main() {
  test('lat range test', () {
    expect(() => UTM.fromLatLon(lat: -81, lon: 0), throwsRangeError);
    expect(UTM.fromLatLon(lat: -80, lon: 0), TypeMatcher<UtmCoordinate>());
    expect(UTM.fromLatLon(lat: 84, lon: 0), TypeMatcher<UtmCoordinate>());
    expect(() => UTM.fromLatLon(lat: 85, lon: 0), throwsRangeError);
  });
  test('lat range test', () {
    expect(() => UTM.fromLatLon(lat: 0, lon: -181), throwsRangeError);
    expect(UTM.fromLatLon(lat: 0, lon: -180), TypeMatcher<UtmCoordinate>());
    expect(UTM.fromLatLon(lat: 0, lon: 180), TypeMatcher<UtmCoordinate>());
    expect(() => UTM.fromLatLon(lat: 0, lon: 181), throwsRangeError);
  });
  test('zone number range test', () {
    expect(
        () => UTM.fromUtm(
            easting: 1, northing: 1, zoneNumber: 0, zoneLetter: 'S'),
        throwsRangeError);
    expect(UTM.fromUtm(easting: 1, northing: 1, zoneNumber: 1, zoneLetter: 'S'),
        TypeMatcher<UtmCoordinate>());
    expect(
        UTM.fromUtm(easting: 1, northing: 1, zoneNumber: 60, zoneLetter: 'S'),
        TypeMatcher<UtmCoordinate>());
    expect(
        () => UTM.fromUtm(
            easting: 1, northing: 1, zoneNumber: 61, zoneLetter: 'S'),
        throwsRangeError);
  });
  test('invalid zone letter', () {
    expect(
        () => UTM.fromUtm(
            easting: -1, northing: 1, zoneNumber: 1, zoneLetter: 'S'),
        throwsRangeError);
  });
  test('invalid zone letter', () {
    expect(
        () => UTM.fromUtm(
            easting: 1, northing: 1, zoneNumber: 1, zoneLetter: 'A'),
        throwsArgumentError);
    expect(
        () => UTM.fromUtm(
            easting: 1, northing: 1, zoneNumber: 1, zoneLetter: 'Z'),
        throwsArgumentError);
  });
}


lib/src/utm_base.dart
// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
import 'package:utm/src/constants.dart';

/// This is a lat lon and UTM bidirectional converter
import 'package:utm/src/converter.dart';
import 'package:utm/src/validator.dart';

/// UTM
class UTM {
  /// create [UtmCoordinate] from UTM coordinates.
  /// [easting] easting value. x-coordinate
  /// [northing] northing value. y-coordinate
  /// [zoneNumber] number of UTM zone. 1 to 60
  /// [zoneLetter] String value of UTM zone. CDEFGHJKLMNPQRSTUVWXX
  /// [type] type of Geodetic System. Default is WGS84. see [GeodeticSystemType]
  static UtmCoordinate fromUtm({
    required double easting,
    required double northing,
    required int zoneNumber,
    required String zoneLetter,
    GeodeticSystemType type = GeodeticSystemType.wgs84,
  }) {
    validateUtmZone(
        easting: easting,
        northing: northing,
        zoneLetter: zoneLetter,
        zoneNumber: zoneNumber);
    return UtmConverter(type)
        .utmToLatLon(easting, northing, zoneNumber, zoneLetter);
  }

  /// create [UtmCoordinate] from latitude & longitude
  /// [lat] latitude between -80 and 84 deg
  /// [lon] longitude between -180 and 180 deg
  /// [type] type of Geodetic System. Default is WGS84. see [GeodeticSystemType]
  static UtmCoordinate fromLatLon({
    required double lat,
    required double lon,
    GeodeticSystemType type = GeodeticSystemType.wgs84,
  }) {
    validateRangeOfLatLon(lat: lat, lon: lon);
    return UtmConverter(type).latlonToUtm(lat, lon);
  }
}


lib/src/constants.dart
/// type of geodetic system
enum GeodeticSystemType {
  /// WGS84 R=6378137 e=0.00669438
  wgs84,

  /// GRS80 R=6378137 e=0.00669438
  grs80,

  /// bessel R=6377397.155 e=0.00667437
  bessel
}

/// zone letters
final zoneLetters = 'CDEFGHJKLMNPQRSTUVWXX';

///
class UtmCoordinate {
  /// Latitude
  final double lat;

  /// Longitude
  final double lon;

  /// the eastward-measured distance. (x-coordinate)
  final double easting;

  /// northward-measured distance. (y-coordinate)
  final double northing;

  /// zone number
  final int zoneNumber;

  /// zone letter
  final String zoneLetter;

  /// zone ex) 60S 50H
  String get zone => zoneNumber.toString() + zoneLetter;

  /// constructor
  const UtmCoordinate(this.lat, this.lon, this.easting, this.northing,
      this.zoneNumber, this.zoneLetter);
}


lib/src/validator.dart
import 'package:utm/src/constants.dart';

const _minLat = -80;
const _maxLat = 84;
const _minLon = -180;
const _maxLon = 180;
const _minZone = 1;
const _maxZone = 60;

/// validate range of lat & lon
void validateRangeOfLatLon({
  required double lat,
  required double lon,
}) {
  if (lat < _minLat || _maxLat < lat) {
    throw RangeError.range(lat, _minLat, _maxLat);
  }
  if (lon < _minLon || _maxLon < lon) {
    throw RangeError.range(lon, _minLon, _maxLon);
  }
}

/// validate utm zone
void validateUtmZone({
  required double easting,
  required double northing,
  required int zoneNumber,
  required String zoneLetter,
}) {
  if (easting < 0) {
    throw RangeError.value(easting);
  }
  if (northing < 0) {
    throw RangeError.value(easting);
  }
  if (zoneNumber < _minZone || _maxZone < zoneNumber) {
    throw RangeError.range(zoneNumber, _minZone, _maxZone);
  }
  if (!zoneLetters.contains(zoneLetter)) {
    throw ArgumentError.value(zoneLetter);
  }
}


lib/src/converter.dart
import 'dart:math' as math;
import 'package:angles/angles.dart';
import 'package:utm/src/constants.dart';

class _GeoSysParam {
  final double r;
  final double e;

  _GeoSysParam(this.r, this.e);
}

final Map<GeodeticSystemType, _GeoSysParam> _geoSysMap = {
  GeodeticSystemType.wgs84: _GeoSysParam(6378137, 0.00669438),
  GeodeticSystemType.grs80: _GeoSysParam(6378137, 0.00669438),
  GeodeticSystemType.bessel: _GeoSysParam(6377397.155, 0.00667437),
};

/// UTM bidirectional converter
class UtmConverter {
  final _k0 = 0.9996;
  late double _r;
  late double _e;
  late double _e2;
  late double _e3;
  late double _eP2;

  late double _sqrtE;
  late double __e;
  late double __e2;
  late double __e3;
  late double __e4;
  late double __e5;

  late double _m1;
  late double _m2;
  late double _m3;
  late double _m4;

  late double _p2;
  late double _p3;
  late double _p4;
  late double _p5;

  /// constructor
  UtmConverter(GeodeticSystemType type) {
    final param = _geoSysMap[type];
    _r = param!.r;
    _e = param.e;
    _e2 = _e * _e;
    _e3 = _e2 * _e;
    _eP2 = _e / (1.0 - _e);

    _sqrtE = math.sqrt(1 - _e);
    __e = (1 - _sqrtE) / (1 + _sqrtE);
    __e2 = __e * __e;
    __e3 = __e2 * __e;
    __e4 = __e3 * __e;
    __e5 = __e4 * __e;

    _m1 = (1 - _e / 4 - 3 * _e2 / 64 - 5 * _e3 / 256);
    _m2 = (3 * _e / 8 + 3 * _e2 / 32 + 45 * _e3 / 1024);
    _m3 = (15 * _e2 / 256 + 45 * _e3 / 1024);
    _m4 = (35 * _e3 / 3072);

    _p2 = (3 / 2 * __e - 27 / 32 * __e3 + 269 / 512 * __e5);
    _p3 = (21 / 16 * __e2 - 55 / 32 * __e4);
    _p4 = (151 / 96 * __e3 - 417 / 128 * __e5);
    _p5 = (1097 / 512 * __e4);
  }

  /// convert to lat&lon from UTM
  UtmCoordinate utmToLatLon(
      double easting, double northing, int zoneNumber, String? zoneLetter) {
    final northern = zoneLetter != null &&
        zoneLetter.toUpperCase().codeUnitAt(0) >= 'N'.codeUnitAt(0);
    final x = easting - 500000;
    final y = northern ? northing : northing - 10000000;

    final m = y / _k0;
    final mu = m / (_r * _m1);

    final pRad = (mu +
        _p2 * math.sin(2 * mu) +
        _p3 * math.sin(4 * mu) +
        _p4 * math.sin(6 * mu) +
        _p5 * math.sin(8 * mu));

    final pSin = math.sin(pRad);
    final pSin2 = pSin * pSin;

    final pCos = math.cos(pRad);

    final pTan = pSin / pCos;
    final pTan2 = pTan * pTan;
    final pTan4 = pTan2 * pTan2;

    final epSin = 1 - _e * pSin2;
    final epSinSqrt = math.sqrt(1 - _e * pSin2);

    final n = _r / epSinSqrt;
    final r = (1 - _e) / epSin;

    final c = __e * math.pow(pCos, 2);
    final c2 = c * c;

    final d = x / (n * _k0);
    final d2 = d * d;
    final d3 = d2 * d;
    final d4 = d3 * d;
    final d5 = d4 * d;
    final d6 = d5 * d;

    final lat = (pRad -
        (pTan / r) *
            (d2 / 2 - d4 / 24 * (5 + 3 * pTan2 + 10 * c - 4 * c2 - 9 * _eP2)) +
        d6 /
            720 *
            (61 + 90 * pTan2 + 298 * c + 45 * pTan4 - 252 * _eP2 - 3 * c2));

    final lon = (d -
            d3 / 6 * (1 + 2 * pTan2 + c) +
            d5 /
                120 *
                (5 - 2 * c + 28 * pTan2 - 3 * c2 + 8 * _eP2 + 24 * pTan4)) /
        pCos;

    return UtmCoordinate(
      Angle.radians(lat).degrees,
      Angle.radians(lon).degrees + _zoneNumber2CentralLon(zoneNumber),
      easting,
      northing,
      zoneNumber,
      zoneLetter!,
    );
  }

  /// convert to UTM from lat&lon
  UtmCoordinate latlonToUtm(double lat, double lon) {
    final latRad = Angle.degrees(lat).radians;
    final latSin = math.sin(latRad);
    final latCos = math.cos(latRad);
    final latTan = latSin / latCos;
    final latTan2 = latTan * latTan;
    final latTan4 = latTan2 * latTan2;
    final zoneNumber = _latlon2zoneNumber(lat, lon);
    final lonRad = Angle.degrees(lon).radians;
    final centralLon = _zoneNumber2CentralLon(zoneNumber);
    final centralLonRad = Angle.degrees(centralLon).radians;

    final n = _r / math.sqrt(1 - _e * math.pow(latSin, 2));
    final c = _eP2 * math.pow(latCos, 2);
    final a = latCos * (lonRad - centralLonRad);
    final a2 = a * a;
    final a3 = a2 * a;
    final a4 = a3 * a;
    final a5 = a4 * a;
    final a6 = a5 * a;
    final m = _r *
        (_m1 * latRad -
            _m2 * math.sin(2 * latRad) +
            _m3 * math.sin(4 * latRad) -
            _m4 * math.sin(6 * latRad));
    final easting = _k0 *
            n *
            (a +
                a3 / 6 * (1 - latTan2 + c) +
                a5 / 120 * (5 - 18 * latTan2 + latTan4 + 72 * c - 58 * _eP2)) +
        500000;
    final offset = lat >= 0 ? 0 : 10000000;
    final northing = _k0 *
            (m +
                n *
                    latTan *
                    (a2 / 2 +
                        a4 / 24 * (5 - latTan2 + 9 * c + 4 * math.pow(c, 2)) +
                        a6 /
                            720 *
                            (61 -
                                58 * latTan2 +
                                latTan4 +
                                600 * c -
                                330 * _eP2))) +
        offset;
    return UtmCoordinate(
        lat, lon, easting, northing, zoneNumber, _lat2zoneLetter(lat));
  }

  String _lat2zoneLetter(double lat) {
    if (-80 <= lat && lat <= 84) {
      return zoneLetters[(lat + 80).toInt() >> 3];
    } else {
      return '';
    }
  }
}

int _latlon2zoneNumber(double lat, double lon) {
  if (56 <= lat && lat < 64 && 3 <= lon && lon < 12) {
    return 32;
  }
  if (72 <= lat && lat <= 84 && lon >= 0) {
    if (lon < 9) {
      return 31;
    } else if (lon < 21) {
      return 33;
    } else if (lon < 33) {
      return 35;
    } else if (lon < 42) {
      return 37;
    }
  }
  return (((lon + 180) / 6) + 1).floor();
}

double _zoneNumber2CentralLon(int zoneNumber) {
  return ((zoneNumber - 1) * 6.0 - 180.0 + 3.0);
}
